File: Runtime_ Associated/Runtime_ Associated/ViewController.m
//
//  ViewController.m
//  Runtime_ Associated
//
//  Created by Kerry on 16/1/20.
//  Copyright © 2016年 DKT. All rights reserved.
//

#import "ViewController.h"

#import "UIViewController+Associated.h"

__weak NSString *string_weak_assign = nil;
__weak NSString *string_weak_retain = nil;
__weak NSString *string_weak_copy   = nil;

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.associatedObject_assign = [NSString stringWithFormat:@"dengketao1"];
    self.associatedObject_retain = [NSString stringWithFormat:@"dengketao2"];
    self.associatedObject_copy   = [NSString stringWithFormat:@"dengketao3"];
//    self.associatedObject_bool = NO;

    string_weak_assign = self.associatedObject_assign;
    string_weak_retain = self.associatedObject_retain;
    string_weak_copy   = self.associatedObject_copy;
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    NSLog(@"self.associatedObject_assign: %@", self.associatedObject_assign); 
    NSLog(@"self.associatedObject_retain: %@", self.associatedObject_retain);
    NSLog(@"self.associatedObject_copy:%@", self.associatedObject_copy);
//    NSLog(@"associatedObject_bool:%d", self.associatedObject_bool);
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: SwiftLearning/Swift_ Closures/Swift_ Closures/ViewController.swift
//
//  ViewController.swift
//  Swift_ Closures
//
//  Created by Kerry on 16/3/4.
//  Copyright © 2016年 DKT. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let arr = [1, 2, 3, 4]
        let arr1 = Getlist(arr) { (s:Int) -> Bool in
            s > 2
        }
        print(arr1)
        
        /// 省略参数类型和括号
        let arr2 = Getlist(arr, pre: {s in return s > 2})
        print(arr2)
        
        /// 省略参数类型、括号、return关键字
        let arr3 = Getlist(arr, pre: {s in s > 2})
        print(arr3)
        
        /// 参数名缩写
        let arr4 = Getlist(arr, pre: {$0>2})
        print(arr4)
        
        Fuction(1) { () -> () in
            print("hello")
        }
    }
    
    func Getlist(arr:[Int], pre:(Int) -> Bool) -> [Int]{
        
        var tempArr = [Int]()
        for temp in arr{
            if pre(temp){
                tempArr.append(temp)
            }
        }
        return tempArr
    }
    
    func Fuction(s:Int, closure:()->()){
        print(s)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}
